scripts/prepare_heat_demand.py

```diff
@@ -0,0 +1,367 @@
+                years.append(year_in_array)
+
+    years = sorted(years)
+    return years
+
+
+def get_holidays(path):
+    """
+    This function determines all holidays of a given region in a given year
+
+    Parameters
+    ----------
+    path : str
+        input path
+    Returns
+    -------
+    holidays : dict
+        dictionary with holidays
+
+    """
+    # Read all national holidays per state
+    all_holidays = pd.read_csv(path)
+    holidays = {}
+
+    # Get holidays in region
+    for row in all_holidays.iterrows():
+        if year == row[1]["year"] and region in row[1]["region"]:
+            holidays[
+                datetime.date(row[1]["year"], row[1]["month"], row[1]["day"])
+            ] = row[1]["holiday"]
+
+    return holidays
+
+
+def check_central_decentral(demands, value, consumer, heat_type):
+    """
+    This function checks whether
+    Parameters
+    ----------
+    demands : pd.DataFrame
+         DataFrame that contains yearly demands
+    value : float
+         Value of the yearly demand
+    consumer : str
+         Name of the consumer (eg.: ghd, efh, mfh)
+    heat_type : str
+         Name of heat type (eg.: central, decentral)
+
+    Returns
+    -------
+    demands : pd.DataFrame
+         Updated DataFrame that contains yearly demands
+    """
+    if consumer + "_heat_" + heat_type in demands.columns:
+        demands[consumer + "_heat_" + heat_type] = np.sum(
+            demands[consumer + "_heat_" + heat_type], [value]
+        )
+    else:
+        demands[consumer + "_heat_" + heat_type] = [value]
+    return demands
+
+
+def get_heat_demand(path, scenario):
+    """
+    This function returns ghd and hh demands together with their unit of a given region and a
+    given scenario
+
+    Parameters
+    ----------
+    path : str
+        input path
+    scenario : str
+        scenario e.g. "base"
+
+    Returns
+    -------
+    demand : DataFrame
+        Dataframe with total yearly demand of central and decentral heat per consumer
+        (eg.: ghd, efh, mfh)
+    demand_unit : str
+        unit of total demands
+
+    """
+    # Read state heat demands of ghd and hh sectors
+    sc = dp.load_b3_scalars(path)  # TODO: Add ', sep=";"' if from schema
+    consumers = ["ghd", "efh", "mfh"]
+    heat_types = ["central", "decentral"]
+    demands = pd.DataFrame()
+    units = []
+
+    for row in sc.iterrows():
+        # Get heat demands in region
+        if (
+            "heat" in row[1]["carrier"]
+            and "demand" in row[1]["tech"]
+            and row[1]["scenario"] == scenario
+            and row[1]["region"] == region
+        ):
+            for consumer in consumers:
+                for heat_type in heat_types:
+                    if (
+                        consumer in row[1]["name"]
+                        and heat_type in row[1]["carrier"]
+                        and "de" + heat_type not in row[1]["carrier"]
+                    ):
```

Comment:
This nested loop is hard to understand. You are checking something, but only for a subset of the scalars, right? You could filter the DataFrame first (`df = df.loc[<some condition>]`) and then run the check on that data.

Reply:
I cleaned it up with commit https://github.com/rl-institut/oemof-B3/pull/76/commits/01d7f745fee06f78ce21fe1e25b2f330cd535e1e.